### DISCUSSION TITLE
Easier Foundry Coverage

### DISCUSSION THREAD

**Discussion by ChefAharoni** [Cyfrin/foundry-full-course-cu#450] [General]
Hey all,

I've found [this post](https://medium.com/@rohanzarathustra/forge-coverage-overview-744d967e112f) on detailed coverage, most of it Patrick have already covered; But the cool new thing I've found is this VSCode extension that shows you inside your code what isn't covered! 
Seems to me much easier than reading from a report.

The extension is called **[Coverage Gutters](https://marketplace.visualstudio.com/items?itemName=ryanluker.vscode-coverage-gutters)**, shown in this screenshot:
<img width="651" alt="image" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/4399057/218b6dce-e651-4710-9911-7921417a6650">

All you have to do it install the extension, run `forge coverage --report lcov;`, and then from the command pallet toggle `Display Coverage` _(It seems to me you have to do this in every file)_.

I think it really helps in looking at which code isn't included in the tests.
***
<br>
Also, I wanted to suggest you do this on a makefile so you won't have to run these commands every time.
I keep my coverage reports in a directory located in the test dir.
<br>

```make
.PHONY: report
# Report shows each line of code what exactly isn't covered.
report:
	forge coverage --report debug > test/coverage/CoverageReportDebug.txt;
	echo "Coverage debug report generated in test/coverage/CoverageReportDebug.txt"

# Reading an lcov.info file is hard, should be converted to html or with the VSCode extension.
lcov:
	forge coverage --report lcov;
	mv lcov.info test/coverage/;
	# genhtml -o report --branch-coverage; # Working only on Ubuntu.
	echo "Coverage lcov report generated in test/coverage/lcov.info";
```

<br>
Happy coding!

**Comment by ghost** [comment]
Cool, thank you!

**Comment by ReverendLovetta** [comment]
Hello everyone, 

After seeing Patrick's way of reading from the coverage file, I also decided to share how to check your coverage via your default web browser using `lcov` (I used to do this when I wrote C code).

For doing this you need to add a new task in your Makefile:

```bash
open_coverage_in_browser: clean
	forge coverage --report lcov --out ./cov/
	genhtml lcov.info --out ./html/
	@$(call open_html,html/index.html)

define open_html
    @case "$$(uname -s)" in \
        Darwin) \
            echo "Opening $(1) on macOS.";\
            open $(1) ;;\
        Linux) \
            echo "Opening $(1) on Linux.";\
            xdg-open $(1) ;;\
        CYGWIN*|MINGW32*|MSYS*|MINGW*) \
            echo "Opening $(1) on Windows.";\
            cmd.exe /c start $(1) ;;\
        *) \
            echo "Unsupported OS." ;;\
    esac
endef
```

And it also good to update your `clean` task:

```bash
# Clean the repo
clean:
	@echo "Cleaning..."
	@forge clean
	@rm -rf cov/
	@rm -rf html/
	@rm -rf lcov.info
	@echo "Cleaning is done!"
```

**Reply by EngrPips** [reply]
Thank you so much for sharing this with us.
